File: LeetCode/Hard/1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree.cpp
class Solution {
public:
    vector<vector<int>> findCriticalAndPseudoCriticalEdges(int n, vector<vector<int>>& edges) {
        for (int i = 0; i < edges.size(); ++i) {
            edges[i].push_back(i);
        }

        std::sort(edges.begin(), edges.end(), [](const std::vector<int>& a, const std::vector<int>& b) {
            return a[2] < b[2];
        });

        int minCost = kruskal(n, edges);
        std::vector<int> criticalEdges, pseudoCriticalEdges;

        for (int i = 0; i < edges.size(); ++i) {
            if (kruskal(n, edges, i) > minCost) {
                criticalEdges.push_back(edges[i][3]);
            } else if (kruskal(n, edges, -1, i) == minCost) {
                pseudoCriticalEdges.push_back(edges[i][3]);
            }
        }

        return { criticalEdges, pseudoCriticalEdges };
    }

    int find(std::vector<int>& parents, int x) {
        if (parents[x] != x) {
            parents[x] = find(parents, parents[x]);
        }
        return parents[x];
    }

    int kruskal(int n, std::vector<std::vector<int>>& edges, int exclude = -1, int include = -1) {
        std::vector<int> parents(n), rank(n, 0);
        for (int i = 0; i < n; ++i) parents[i] = i;

        int cost = 0, numEdges = 0;

        if (include != -1) {
            int a = edges[include][0], b = edges[include][1], weight = edges[include][2];
            parents[find(parents, a)] = find(parents, b);
            cost += weight;
            numEdges += 1;
        }

        for (int i = 0; i < edges.size(); ++i) {
            if (i == exclude) continue;
            int a = edges[i][0], b = edges[i][1], weight = edges[i][2];
            int parentA = find(parents, a), parentB = find(parents, b);
            if (parentA != parentB) {
                if (rank[parentA] < rank[parentB]) {
                    parents[parentA] = parentB;
                } else {
                    parents[parentB] = parentA;
                    if (rank[parentA] == rank[parentB]) {
                        rank[parentA] += 1;
                    }
                }
                cost += weight;
                numEdges += 1;
                if (numEdges == n - 1) break;
            }
        }

        return numEdges == n - 1 ? cost : INT_MAX;
    }

};

File: LeetCode/Hard/1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/1489-find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree.py
class Solution:
    def findCriticalAndPseudoCriticalEdges(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        # Adding index to the edges to keep track of original indices
        for i, edge in enumerate(edges):
            edges[i].append(i)
        # Sorting the edges by weight
        edges.sort(key=lambda x: x[2])

        def find(parents, x):
            if parents[x] != x:
                parents[x] = find(parents, parents[x])
            return parents[x]

        def kruskal(exclude=-1, include=-1):
            parents = [i for i in range(n)]
            rank = [0] * n
            cost = 0
            numEdges = 0
            
            if include != -1:
                a, b, weight, _ = edges[include]
                parents[find(parents, a)] = find(parents, b)
                cost += weight
                numEdges += 1
            
            for i, (a, b, weight, index) in enumerate(edges):
                if i == exclude:
                    continue
                parentA = find(parents, a)
                parentB = find(parents, b)
                if parentA != parentB:
                    if rank[parentA] < rank[parentB]:
                        parents[parentA] = parentB
                    else:
                        parents[parentB] = parentA
                        if rank[parentA] == rank[parentB]:
                            rank[parentA] += 1
                    cost += weight
                    numEdges += 1
                    if numEdges == n - 1:
                        break
            return cost if numEdges == n - 1 else float('inf')

        minCost = kruskal()
        criticalEdges = []
        pseudoCriticalEdges = []

        for i, edge in enumerate(edges):
            if kruskal(exclude=i) > minCost:
                criticalEdges.append(edge[3])
            elif kruskal(include=i) == minCost:
                pseudoCriticalEdges.append(edge[3])

        return [criticalEdges, pseudoCriticalEdges]

